Add documentation for graph visualization using Graphviz

YueRen opened the pull request:
Related to https://github.com/oscar-system/Oscar.jl/issues/5166

- added a half sentence on Graphviz to the `visualize(::Graph)` docstring
- added a paragraph on reconfiguring polymake to find Graphviz to the docu

YueRen committed:
Combinatorics: new docu for graph visualization via Graphviz

lgoettgens commented on docs/src/Combinatorics/graphs.md:
````diff
@@ -110,6 +110,15 @@ src(e::Edge)
+graphs using Graphviz, use `backend=:graphviz`. It requires that both Graphviz and a postscript
+viewer are installed. In case Graphviz is installed and OSCAR is unable to find it (e.g., because
+Graphviz was installed after OSCAR), please run
+```
````
````suggestion
```julia
````

lgoettgens commented on src/Combinatorics/Graphs/functions.jl:
```diff
@@ -1325,12 +1325,20 @@ end
-`white`, `purple`, `cyan`, `darkolivegreen`, `indianred`, `plum1`, `red`, `lightslategrey`, `yellow`, `orange`, `salmon1`, `azure`, `green`, `gray`, `midnightblue`, `pink`, `magenta`, `blue`, `lavenderblush`, `chocolate1`, `lightgreen`, `black`.
+The `backend` keyword argument allows the user to pick between a Three.js visualization by default,
+or passing `:tikz` for a TikZ visualization, or passing `:graphviz` for a
+[Graphviz](https://graphviz.org/) visualization (requires Graphviz, see documentation).
```
```suggestion
[Graphviz](https://graphviz.org/) visualization (requires Graphviz, see [Visualization using Graphviz](@ref)).
```

YueRen committed:
Update docs/src/Combinatorics/graphs.md

Co-authored-by: Lars Göttgens <lars.goettgens@gmail.com>

YueRen committed:
Update src/Combinatorics/Graphs/functions.jl

Co-authored-by: Lars Göttgens <lars.goettgens@gmail.com>

benlorenz commented:
There is also a `Graphviz_jll` package that we could add and have it available for everyone. That would avoid having users run this polymake shell command. But unfortunately it has quite a few transitive dependencies:
```
    Updating `/tmp/jl_VqdbKE/Project.toml`
  [3c863552] + Graphviz_jll v2.50.0+1
    Updating `/tmp/jl_VqdbKE/Manifest.toml`
  [692b3bcd] + JLLWrappers v1.7.1
  [21216c6a] + Preferences v1.5.0
  [6e34b625] + Bzip2_jll v1.0.9+0
  [83423d85] + Cairo_jll v1.18.5+0
  [2e619515] + Expat_jll v2.6.5+0
  [a3f928ae] + Fontconfig_jll v2.16.0+0
  [d7e528f0] + FreeType2_jll v2.13.4+0
  [559328eb] + FriBidi_jll v1.0.17+0
  [b0724c58] + GettextRuntime_jll v0.22.4+0
  [7746bdde] + Glib_jll v2.84.3+0
  [3b182d85] + Graphite2_jll v1.3.15+0
  [3c863552] + Graphviz_jll v2.50.0+1
  [2e76f6c2] + HarfBuzz_jll v8.5.1+0
  [1d63c593] + LLVMOpenMP_jll v18.1.8+0
  [dd4b983a] + LZO_jll v2.10.3+0
  [e9f186c6] + Libffi_jll v3.4.7+0
  [94ce4f54] + Libiconv_jll v1.18.0+0
  [4b2f31a3] + Libmount_jll v2.41.0+0
  [38a345b3] + Libuuid_jll v2.41.0+0
  [36c8627f] + Pango_jll v1.56.3+0
⌅ [30392449] + Pixman_jll v0.44.2+0
  [4f6342f7] + Xorg_libX11_jll v1.8.12+0
  [0c0b7dd1] + Xorg_libXau_jll v1.0.13+0
  [a3789734] + Xorg_libXdmcp_jll v1.1.6+0
  [1082639a] + Xorg_libXext_jll v1.3.7+0
  [ea2f1a96] + Xorg_libXrender_jll v0.9.12+0
  [c7cfdc94] + Xorg_libxcb_jll v1.17.1+0
  [c5fb5394] + Xorg_xtrans_jll v1.6.0+0
  [b53b4c65] + libpng_jll v1.6.50+0
```

Maybe we could add it as a weak dependency to `Polymake.jl` and add the corresponding configuration step to the weakdep code?

benlorenz commented:
I have created a draft with such a Graphviz weakdep here: https://github.com/oscar-system/Polymake.jl/pull/530
If someone wants to use it they can just make sure `Graphviz_jll` is in the environment and loaded before Oscar/Polymake, then it will be automatically configured to be active and set the correct paths.

fingolfin commented on src/Combinatorics/Graphs/functions.jl:
```diff
@@ -1325,12 +1325,20 @@ end
+or passing `:tikz` for a TikZ visualization, or passing `:graphviz` for a
+[Graphviz](https://graphviz.org/) visualization (requires Graphviz, see [Visualization using Graphviz](@ref)).
+
+The`filename` keyword argument will write visualization code to the `filename` location, this will
```
```suggestion
The `filename` keyword argument will write visualization code to the `filename` location, this will
```
or even
```suggestion
Providing the `filename` keyword argument will write visualization code to the `filename` location, this will
```

YueRen committed:
Update src/Combinatorics/Graphs/functions.jl

Co-authored-by: Max Horn <max@quendi.de>

lgoettgens commented:
Now with https://github.com/oscar-system/Polymake.jl/pull/530 merged and available, the paragraph should no longer recommend using `Polymake.shell` but instead load `Graphviz_jll`

YueRen commented:
If using `Graphviz` is supposed to be automatic, then there seems to be some problem.  The following code gives me the same straight line visualization as discussed [here](https://github.com/oscar-system/Oscar.jl/issues/5166): 

```
using Graphviz_jll # making sure to load it before Oscar
using Oscar
G = graph_from_edges([[1,2], [1,3], [2,3], [2,4], [3,4], [3,5], [4,5], [4,6], [5,6]])
visualize(G)
```

My `Polymake.jl` version should be correct, here the relevant snippet from my `Manifest.toml`:
```
[[deps.Polymake]]
deps = ["AbstractAlgebra", "BinaryWrappers", "CxxWrap", "FileWatching", "JSON", "Mongoc", "NetworkOptions", "Ninja_jll", "Perl_jll", "Pkg", "REPL", "Scratch", "SparseArrays", "TOPCOM_jll", "lib4ti2_jll", "libpolymake_julia_jll", "polymake_jll", "polymake_oscarnumber_jll"]
git-tree-sha1 = "e7144a6393cb4fdc13c1963273a989977e7babd0"
uuid = "d720cf60-89b5-51f5-aff5-213f193123e7"
version = "0.13.1"
weakdeps = ["Graphviz_jll"]

    [deps.Polymake.extensions]
    GraphvizExt = "Graphviz_jll"
```

edit: Not sure whether it is supposed to help, but I tried running the commands in `__init__()` of `GraphvizExt.jl` manually and it changed nothing on my system:

```
julia> binpath = Polymake.@generate_wrappers(Graphviz_jll)
"/home/ren/.julia/scratchspaces/3c863552-8265-54e4-a6dc-903eb78fde85/binarywrappers_v1.11/bin/"

julia> Polymake.shell_execute(raw"""application("graph")->reconfigure("graphviz.rules");""")
(stdout = "", stderr = "")

julia> Polymake.shell_execute("""\$Graphviz::dot = "$binpath/dot";""")
(stdout = "", stderr = "")

julia> Polymake.shell_execute(raw"""application("graph")->reconfigure("graphviz.rules");""")
(stdout = "", stderr = "")

julia> visualize(G) # same visualization as before
```

YueRen committed:
Combinatorics: updated docs and docstring for graph visualization

YueRen committed:
Combinatorics: changed graph visualization docstring

YueRen committed:
Combinatorics: changed graph visualization docstring

YueRen commented:
I changed the docs accordingly, the new docstring now reads:

```
@doc raw"""
    visualize(G::Graph{<:Union{Polymake.Directed, Polymake.Undirected}}; backend::Symbol=:threejs, filename::Union{Nothing, String}=nothing, kwargs...)

Visualize graph `G`.

The `backend` keyword argument allows the user to pick between
- `:threejs`: a Three.js visualization (default),
- `:tikz`: a TikZ visualization, or
- `:graphviz`: a [Graphviz](https://graphviz.org/) visualization.
The Graphviz visualization requires a postscript reader and loading the julia
package `Graphviz_jll` before OSCAR.

If `filename` is specified, OSCAR will write visualization code to `filename`
(html for `:threejs`, TikZ code for `:tikz`, dot for `:graphviz`).

If `G` has a labeling `:color` (see [`label!`](@ref)) then the visualization
will use these colors to color the graph.  Possible color labelings include RGB
values of the form `"255 0 255"` or `"#ff00ff"`, as well as the following named
colors as strings: `polymakeorange`, `polymakegreen`, `white`, `purple`, `cyan`,
`darkolivegreen`, `indianred`, `plum1`, `red`, `lightslategrey`, `yellow`,
`orange`, `salmon1`, `azure`, `green`, `gray`, `midnightblue`, `pink`,
`magenta`, `blue`, `lavenderblush`, `chocolate1`, `lightgreen`, `black`.

"""
```

I have also removed the "Visualization with Graphviz" section in the documentation that was added temporarily, because setting up Graphviz is now so easy that it can be fully described in the docstring (see above).

benlorenz reviewed (approved):
thanks!

YueRen commented:
I'm not sure why some of the tests are failing.  This PR mainly only changes the non-example part of a docstring.  But it should be ready to merge otherwise.

benlorenz commented:
> I'm not sure why some of the tests are failing. This PR mainly only changes the non-example part of a docstring. But it should be ready to merge otherwise.

Nightly is currently failing on all branches, we probably need rebuilds of `libjulia_jll` and related packages to fix this.
